src/functions.jl

```diff
@@ -6,11 +6,37 @@ function MOI.VectorAffineFunction{T}(f::MOI.VectorOfVariables) where T
     MOI.VectorAffineFunction(collect(1:n), f.variables, ones(T, n), zeros(T, n))
 end
 
+# Implements iterator interface
+struct ScalarFunctionIterator{F<:MOI.AbstractVectorFunction}
+    f::F
+end
+eachscalar(f::MOI.AbstractVectorFunction) = ScalarFunctionIterator(f)
+Base.start(it::ScalarFunctionIterator) = 1
+Base.done(it::ScalarFunctionIterator, state) = state > length(it)
+Base.next(it::ScalarFunctionIterator, state) = (it.f[state], state+1)
+Base.length(it::ScalarFunctionIterator) = length(it.f)
+Base.eltype(it::ScalarFunctionIterator) = eltype(it.f)
+
 # Define getindex for Vector functions
+Base.length(f::MOI.VectorOfVariables) = length(f.variables)
+Base.length(f::Union{MOI.VectorAffineFunction, MOI.VectorQuadraticFunction}) = length(f.constant)
+Base.endof(f::MOI.AbstractVectorFunction) = length(f)
+Base.eltype(f::MOI.VectorOfVariables) = MOI.SingleVariable
+Base.eltype(f::MOI.VectorAffineFunction{T}) where T = MOI.ScalarAffineFunction{T}
+Base.eltype(f::MOI.VectorQuadraticFunction{T}) where T = MOI.ScalarQuadraticFunction{T}
+
+Base.getindex(f::MOI.VectorOfVariables, i::Integer) = MOI.SingleVariable(f.variables[i])
 function Base.getindex(f::MOI.VectorAffineFunction, i::Integer)
     I = find(oi -> oi == i, f.outputindex)
     MOI.ScalarAffineFunction(f.variables[I], f.coefficients[I], f.constant[i])
 end
+function Base.getindex(f::MOI.VectorQuadraticFunction, i::Integer)
```

Comment:
Shouldn't `getindex` apply to `ScalarFunctionIterator`, not the MOI function directly?

Reply:
Ah ok, I misunderstood. That is changed now.